server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java

```diff
@@ -0,0 +1,362 @@
+    }
+
+    private final IndexSearcher searcher;
+    private final String key;
+    private final Q query;
+    /**
+     * The weight for the query or {@code null} if we haven't built it. Use
+     * {@link #weight()} to build it when needed.
+     */
+    private Weight weight;
+    /**
+     * Scorer for each segment or {@code null} if we haven't built the scorer.
+     * Use {@link #bulkScorer(LeafReaderContext, Runnable)} to build the scorer
+     * when needed.
+     */
+    private BulkScorer[] bulkScorers;
+    /**
+     * The number of scorers we prepared just to estimate the cost of counting
+     * documents. For some queries preparing the scorers is very slow so its
+     * nice to know how many we built. Exposed by profiling.
+     */
+    private int scorersPreparedWhileEstimatingCost;
+
+    private QueryToFilterAdapter(IndexSearcher searcher, String key, Q query) {
+        this.searcher = searcher;
+        this.key = key;
+        this.query = query;
+    }
+
+    /**
+     * The query we're adapting.
+     * <p>
+     * Subclasses should use this to fetch the query when making query
+     * specific optimizations.
+     */
+    Q query() {
+        return query;
+    }
+
+    /**
+     * Key for this filter.
+     */
+    public final String key() {
+        return key;
+    }
+
+    /**
+     * Searcher that this filter is targeting.
+     */
+    protected final IndexSearcher searcher() {
+        return searcher;
+    }
+
+    /**
+     * Would using index metadata like {@link IndexReader#docFreq}
+     * or {@link IndexReader#maxDoc} to count the number of matching documents
+     * produce the same answer as collecting the results with a sequence like
+     * {@code searcher.collect(counter); return counter.readAndReset();}?
+     */
+    protected final boolean countCanUseMetadata(FiltersAggregator.Counter counter, Bits live) {
+        if (live != null) {
+            /*
+             * We can only use metadata if all of the documents in the reader
+             * are visible. This is done by returning a null `live` bits. The
+             * name `live` is traditional because most of the time a non-null
+             * `live` bits means that there are deleted documents. But `live`
+             * might also be non-null if document level security is enabled.
+             */
+            return false;
+        }
+        /*
+         * We can only use metadata if we're not using the special docCount
+         * field. Otherwise we wouldn't know how many documents each lucene
+         * document represents.
+         */
+        return counter.docCount.alwaysOne();
+    }
+
+    /**
+     * Make a filter that matches this filter and the provided query.
+     * <p>
+     * Note: This method rewrites the query against the {@link IndexSearcher}.
+     */
+    QueryToFilterAdapter<?> union(Query extraQuery) throws IOException {
+        extraQuery = searcher().rewrite(extraQuery);
```

Comment:
This method seems like a limited duplication of logic of boolean query rewrite. What's the main reason for not just wrapping this into boolean with 2 must closes and allow it to figure out optimization?

Reply:
BooleanQuery doesn't know how to merge point range queries and that speeds up when we rewrite `date_histogram` into filters by a factor of about 2. Its something @iverase and I've talked about some. And I've talked about it with @romseygeek. I think at some point Lucene'll be able to do more of this sort of thing, but until then, here we are.

Reply:
Makes sense. I think a TODO here with a link to the issue would help future readers.